pystiche/papers/ulyanov_et_al_2016/core.py

```diff
@@ -0,0 +1,192 @@
+) -> ExponentialLR:
+    if impl_params:
+        lr_scheduler = ExponentialLR(optimizer, 0.8)
+    else:
+        lr_scheduler = DelayedExponentialLR(optimizer, 0.7, 5)
+    return lr_scheduler
+
+
+def ulyanov_et_al_2016_training(
+    content_image_loader: DataLoader,
+    style: Union[str, torch.Tensor],
+    impl_params: bool = True,
+    instance_norm: bool = True,
+    stylization: bool = True,
+    transformer: Optional[ulyanov_et_al_2016_transformer] = None,
+    criterion: Optional[UlyanovEtAl2016PerceptualLoss] = None,
+    lr_scheduler: Optional[ulyanov_et_al_2018_lr_scheduler] = None,
+    num_epochs: Optional[int] = None,
+    get_optimizer: Optional[
+        Callable[[ulyanov_et_al_2016_transformer], Optimizer]
+    ] = None,
+    quiet: bool = False,
+    logger: Optional[OptimLogger] = None,
+    log_fn: Optional[
+        Callable[[int, Union[torch.Tensor, pystiche.LossDict], float, float], None]
+    ] = None,
+):
+    if isinstance(style, str):
+        device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+        images = ulyanov_et_al_2016_images(download=False)
+        style_image = images[style].read(device=device)
+    else:
+        style_image = style
+        device = style_image.device
+
+    if transformer is None:
+        transformer = ulyanov_et_al_2016_transformer(
+            impl_params=impl_params,
+            instance_norm=instance_norm,
+            stylization=stylization,
+        )
+        transformer = transformer.train()
+    transformer = transformer.to(device)
+
+    if criterion is None:
+        criterion = ulyanov_et_al_2016_perceptual_loss(
+            impl_params=impl_params,
+            instance_norm=instance_norm,
+            stylization=stylization,
+        )
+        criterion = criterion.eval()
+    criterion = criterion.to(device)
+
+    if lr_scheduler is None:
+        if get_optimizer is None:
+            get_optimizer = ulyanov_et_al_2016_optimizer
+        optimizer = get_optimizer(transformer)
+        lr_scheduler = ulyanov_et_al_2018_lr_scheduler(
+            optimizer, impl_params=impl_params,
+        )
+
+    if num_epochs is None:
+        if impl_params:
+            if instance_norm:
+                num_epochs = 25
+            else:
+                num_epochs = 10 if stylization else 5
+        else:
+            num_epochs = 10
+
+    style_transform = ulyanov_et_al_2016_style_transform(
+        impl_params=impl_params, instance_norm=instance_norm
+    )
+    style_transform = style_transform.to(device)
+    style_image = style_transform(style_image)
+    style_image = batch_up_image(style_image, loader=content_image_loader)
+
+    criterion.set_style_image(style_image)
+
+    def criterion_update_fn(input_image, criterion, preprocessor=None):
+        if hasattr(criterion, "content_loss"):
+            if preprocessor is None:
+                preprocessor = ulyanov_et_al_2016_preprocessor()
+            preprocessor = preprocessor.to(device)
```

Comment:
Do you mean the functions `default_transformer_epoch_optim_loop`  and the functions associated with it should get one more parameter `preprocessor`?

Reply:
Nope. Right now you create `preprocessor` within `criterion_update_fn`. That means that every time `criterion_update_fn` is called, i.e. on every batch, the `preprocessor` is created over and over again. Since the `preprocessor` does not need to change for every batch you can create it on the outside of `criterion_update_fn` and simply use it within.